text/0000-eager-macro-expansion.md

```diff
@@ -0,0 +1,615 @@
+    struct X {
+        my_field_definition_macro!(...)
+    //  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Same with this one.
+    }
+    ```
+
+In these situations, macros need to either re-emit the input macro invocation
+as part of their token output, or simply reject the input. If the proc macro
+needs to inspect the result of the macro call (for instance, to check or edit
+it, or to re-export a hygienic symbol defined in it), the author is currently
+unable to do so.
+
+Giving proc macro authors the ability to handle these situations will allow
+proc macros to 'just work' in more contexts, and without surprising users who
+expect macro calls to interact well with more parts of the language.
+
+As a side note, allowing macro calls in built-in attributes would solve a few
+outstanding issues (see
+[rust-lang/rust#18849](https://github.com/rust-lang/rust/issues/18849) for an
+example).
+
+An older motivation to allow macro calls in attributes was to get
+`#[doc(include_str!("path/to/doc.txt"))]` working, in order to provide an
+ergonomic way to keep documentation outside of Rust source files. This was
+eventually emulated by the accepted [RFC
+1990](https://github.com/rust-lang/rfcs/pull/1990), indicating that macros in
+attributes could be used to solve problems at least important enough to go
+through the RFC process.
+
+## Interpolating macros in output
+
+Macros are currently not allowed in certain syntactic positions. Famously, they
+aren't allowed in identifier position, which makes `concat_idents!` [almost
+useless](https://github.com/rust-lang/rust/issues/29599). If macro authors have
+access to eager expansion, they could eagerly expand `concat_idents!` and
+interpolate the resulting token into their output.
+
+# Guide-level explanation
+
+We expose the following API in the `proc_macro` crate, to allow proc macro
+authors to iteratively expand all macros within a token stream:
+
+```rust
+use proc_macro::{LexError, Diagnostic, TokenStream};
+
+pub struct ExpansionBuilder { /* No public fields. */ }
+pub enum ExpansionError {
+    LexError(LexError),
+    ParseError(Diagnostic),
+    MissingDefinition(Diagnostic),
+    /* What other errors do we want to report?
+       What failures can the user reasonably handle? */
+}
+
+impl ExpansionBuilder {
+    /// Creates a new macro expansion request to iteratively expand all macro
+    /// invocations that occur in `tokens`.
+    ///
+    /// Expansion results will be interpolated within the input stream before
+    /// being returned.
+    ///
+    /// `tokens` should parse as valid Rust -- for instance, as an item or
+    /// expression.
+    pub fn from_tokens(tokens: TokenStream) -> Self;
+
+    /// Sends the expansion request to the compiler, then awaits the results of
+    /// expansion.
+    ///
+    /// The main causes for an expansion not completing right away are:
+    /// - Procedural macros performing IO or complex analysis.
```

Comment:
By 'not completing right away', do you mean 'returing `Poll::Pending`'?

The best practice for an implementation of `Future::poll` is [to return quickly](https://doc.rust-lang.org/std/future/trait.Future.html#runtime-characteristics). This suggests that the `Future` returned by `expand` should deterministically:

1. Return `Poll::Ready` if no macros needed to be expanded. We could possibly expand `macro_rules!` macros, since they don't require running arbitrary user code.
2. Return `Poll::Pending` if we needed to start the expansion of any proc macros. When all of the invoked proc macros complete, `Waker::wake` will (somehow) be called, at which point the future will return `Poll::Ready`

Reply:
The future could also be blocking on a currently-nonexistent decl macro definition (see the section on [why the API is asynchronous](https://github.com/pierzchalski/rfcs/blob/macro-expansion-api/text/0000-eager-macro-expansion.md#why-is-expansion-asynchronous)).

What's the normal/accepted language for saying "the future won't be ready immediately"? Should it say "The main causes for an expansion `Future` not being ready immediately are:"?

Reply:
My main concern is over the language "Procedural macros performing IO or complex analysis" - this sounds like expanding a proc macro that does *not* "perform or complex analysis" could cause `Poll::Ready` to be returned immediately. That is, the implementation of `expand` might wait for some short period of time to see if the invoked proc-macro will compete "quickly enough".